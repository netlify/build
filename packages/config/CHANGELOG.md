# Changelog

### [3.1.2](https://www.github.com/netlify/build/compare/v3.1.1...v3.1.2) (2021-02-02)


### Bug Fixes

* add SITE_ID and SITE_NAME env vars to local builds ([#2239](https://www.github.com/netlify/build/issues/2239)) ([113e4d2](https://www.github.com/netlify/build/commit/113e4d2c48e0264e48f391e5a7d219332d012fab))

### [3.1.1](https://www.github.com/netlify/build/compare/v3.1.0...v3.1.1) (2021-01-29)


### Bug Fixes

* **deps:** update dependency netlify to v6.1.5 ([#2225](https://www.github.com/netlify/build/issues/2225)) ([fec5c83](https://www.github.com/netlify/build/commit/fec5c83c12ebf7572aaf8756fdbec6e9fe1e3699))

## [3.1.0](https://www.github.com/netlify/build/compare/v3.0.5...v3.1.0) (2021-01-27)


### Features

* allow override of API URL ([#2190](https://www.github.com/netlify/build/issues/2190)) ([3e1be70](https://www.github.com/netlify/build/commit/3e1be7057ac9f217405b2cbe15f39e1ecd7496e6))

### [3.0.5](https://www.github.com/netlify/build/compare/v3.0.4...v3.0.5) (2021-01-27)


### Bug Fixes

* **deps:** update dependency netlify to v6.1.4 ([#2219](https://www.github.com/netlify/build/issues/2219)) ([3a75574](https://www.github.com/netlify/build/commit/3a7557400d46388aad42250f6d9751f1617f3cd0))

### [3.0.4](https://www.github.com/netlify/build/compare/v3.0.3...v3.0.4) (2021-01-25)


### Bug Fixes

* **deps:** update dependency netlify to v6.1.3 ([#2210](https://www.github.com/netlify/build/issues/2210)) ([b7c2c40](https://www.github.com/netlify/build/commit/b7c2c4049e9037cf855475543b13e33c32ceb4a6))

### [3.0.3](https://www.github.com/netlify/build/compare/v3.0.2...v3.0.3) (2021-01-25)


### Bug Fixes

* **deps:** update dependency netlify to v6.1.1 ([#2200](https://www.github.com/netlify/build/issues/2200)) ([930805e](https://www.github.com/netlify/build/commit/930805ec70cedbb08d58490c038ea9bcd8ecd35f))

### [3.0.2](https://www.github.com/netlify/build/compare/config-v3.0.1...v3.0.2) (2021-01-22)


### Bug Fixes

* **deps:** update dependency netlify to v6.1.0 ([#2194](https://www.github.com/netlify/build/issues/2194)) ([4b39e9c](https://www.github.com/netlify/build/commit/4b39e9c746c3a8cf1753bd66913883f2adff6ed8))
