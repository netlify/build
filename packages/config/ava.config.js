import baseConfig from '../../ava.base.js'

const config = {
  ...baseConfig,
  workerThreads: false,
  files: ['tests/*.{cjs,mjs,js}', 'tests/**/tests.{cjs,mjs,js}'],
}

export default config
