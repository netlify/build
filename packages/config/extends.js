const { dirname } = require('path')

const deepmerge = require('deepmerge')

// Allow extending configuration with the `extends` property.
// This also allows plugin presets.
const parseExtends = async function({
  config: { extends: subConfigs, ...config },
  configPath,
  context,
  resolveConfig,
}) {
  if (subConfigs === undefined) {
    return config
  }

  const subConfigsA = Array.isArray(subConfigs) ? subConfigs : [subConfigs]

  if (subConfigsA.length === 0) {
    return config
  }

  const cwd = dirname(configPath)
  const subConfigsB = await Promise.all(subConfigsA.map(subConfig => resolveConfig(subConfig, { cwd, context })))
  const configA = subConfigsB.reduce(mergeConfigs, config)
  return configA
}

const mergeConfigs = function(config, subConfig) {
  return deepmerge(subConfig, config)
}

module.exports = { parseExtends }
