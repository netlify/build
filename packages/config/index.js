const configorama = require('configorama')

const { getConfigPath } = require('./path')
const { validateConfig } = require('./validate')
const { normalizeConfig } = require('./normalize')
const { parseExtends } = require('./extends')

const resolveConfig = async function(configFile, { cwd, context } = {}) {
  const configPath = await getConfigPath(configFile, cwd)

  const config = await configorama(configPath, {
    options: { context },
    variableSources: [
      {
        /* Match variables ${secrets:xyz} */
        match: /^secrets:/,
        async resolver() {
          // Call to remote secret store
          return 'shhhhhhh'
        },
      },
      {
        /* Match variables ${context:xyz} */
        match: /^context:/,
        async resolver(varToProcess, { context = 'production' }) {
          const objectPath = varToProcess.replace('context:', '')
          return `\${self:context.${context}.${objectPath}}`
        },
      },
    ],
  })

  validateConfig(config)

  const configA = normalizeConfig(config)
  const configB = await parseExtends({ config: configA, configPath, context, resolveConfig })
  return configB
}

module.exports = resolveConfig
/* Formatting utilities */
module.exports.formatUtils = configorama.format
/* Resolve Netlify config path */
module.exports.getConfigPath = getConfigPath
