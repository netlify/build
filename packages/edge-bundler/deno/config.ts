const [functionURL, collectorURL, bootstrapURL, rawExitCodes] = Deno.args
const exitCodes = JSON.parse(rawExitCodes)

const { Netlify } = await import(bootstrapURL)
globalThis.Netlify = Netlify

let func

try {
  func = await import(functionURL)
} catch (error) {
  console.error(error)

  Deno.exit(exitCodes.ImportError)
}

if (typeof func.default !== 'function') {
  Deno.exit(exitCodes.InvalidDefaultExport)
}

if (func.config === undefined) {
  Deno.exit(exitCodes.NoConfig)
}

if (typeof func.config !== 'object') {
  Deno.exit(exitCodes.InvalidExport)
}

try {
  const result = JSON.stringify(func.config)

  await Deno.writeTextFile(new URL(collectorURL), result)
} catch (error) {
  console.error(error)

  Deno.exit(exitCodes.SerializationError)
}

Deno.exit(exitCodes.Success)
