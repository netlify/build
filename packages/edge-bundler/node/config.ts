import { promises as fs } from 'fs'
import { join } from 'path'
import { pathToFileURL } from 'url'

import tmp from 'tmp-promise'

import { DenoBridge } from './bridge.js'
import { BundleError } from './bundle_error.js'
import { EdgeFunction } from './edge_function.js'
import { ImportMap } from './import_map.js'
import { Logger } from './logger.js'
import { getPackagePath } from './package_json.js'
import { RateLimit } from './rate_limit.js'

enum ConfigExitCode {
  Success = 0,
  UnhandledError = 1,
  ImportError,
  NoConfig,
  InvalidExport,
  SerializationError,
  InvalidDefaultExport,
}

export const enum Cache {
  Off = 'off',
  Manual = 'manual',
}

export type HTTPMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'OPTIONS'

export type Path = `/${string}`

export type OnError = 'fail' | 'bypass' | Path

export const isValidOnError = (value: unknown): value is OnError => {
  if (typeof value === 'undefined') return true
  if (typeof value !== 'string') return false
  return value === 'fail' || value === 'bypass' || value.startsWith('/')
}

interface BaseFunctionConfig {
  cache?: Cache
  onError?: OnError
  name?: string
  generator?: string
  method?: HTTPMethod | HTTPMethod[]
  rateLimit?: RateLimit
}

interface FunctionConfigWithPath extends BaseFunctionConfig {
  path?: Path | Path[]
  excludedPath?: Path | Path[]
}

interface FunctionConfigWithPattern extends BaseFunctionConfig {
  pattern: string | string[]
  excludedPattern?: string | string[]
}

export type FunctionConfig = FunctionConfigWithPath | FunctionConfigWithPattern

export type FunctionConfigWithAllPossibleFields = BaseFunctionConfig &
  Partial<FunctionConfigWithPath & FunctionConfigWithPattern>

const getConfigExtractor = () => {
  const packagePath = getPackagePath()
  const configExtractorPath = join(packagePath, 'deno', 'config.ts')

  return configExtractorPath
}

export const getFunctionConfig = async ({
  func,
  importMap,
  deno,
  log,
}: {
  func: EdgeFunction
  importMap: ImportMap
  deno: DenoBridge
  log: Logger
}) => {
  // The extractor is a Deno script that will import the function and run its
  // `config` export, if one exists.
  const extractorPath = getConfigExtractor()

  // We need to collect the output of the config function, which should be a
  // JSON object. Rather than printing it to stdout, the extractor will write
  // it to a temporary file, which we then read in the Node side. This allows
  // the config function to write to stdout and stderr without that interfering
  // with the extractor.
  const collector = await tmp.file()

  // The extractor will use its exit code to signal different error scenarios,
  // based on the list of exit codes we send as an argument. We then capture
  // the exit code to know exactly what happened and guide people accordingly.
  const { exitCode, stderr, stdout } = await deno.run(
    [
      'run',
      '--allow-env',
      '--allow-net',
      '--allow-read',
      `--allow-write=${collector.path}`,
      '--quiet',
      `--import-map=${importMap.toDataURL()}`,
      '--no-config',
      '--node-modules-dir=false',
      extractorPath,
      pathToFileURL(func.path).href,
      pathToFileURL(collector.path).href,
      JSON.stringify(ConfigExitCode),
    ],
    { rejectOnExitCode: false },
  )

  if (exitCode !== ConfigExitCode.Success) {
    handleConfigError(func, exitCode, stderr, log)

    return {}
  }

  if (stdout !== '') {
    log.user(stdout)
  }

  let collectorData: FunctionConfig = {}

  try {
    const collectorDataJSON = await fs.readFile(collector.path, 'utf8')
    collectorData = JSON.parse(collectorDataJSON) as FunctionConfig
  } catch {
    handleConfigError(func, ConfigExitCode.UnhandledError, stderr, log)
  } finally {
    await collector.cleanup()
  }

  if (!isValidOnError(collectorData.onError)) {
    throw new BundleError(
      new Error(
        `The 'onError' configuration property in edge function at '${func.path}' must be one of 'fail', 'bypass', or a path starting with '/'. Got '${collectorData.onError}'. More on the Edge Functions API at https://ntl.fyi/edge-api.`,
      ),
    )
  }

  return collectorData
}

const handleConfigError = (func: EdgeFunction, exitCode: number, stderr: string, log: Logger) => {
  switch (exitCode) {
    case ConfigExitCode.ImportError:
      log.user(stderr)
      throw new BundleError(
        new Error(
          `Could not load edge function at '${func.path}'. More on the Edge Functions API at https://ntl.fyi/edge-api.`,
        ),
      )

      break

    case ConfigExitCode.NoConfig:
      log.system(`No in-source config found for edge function at '${func.path}'`)

      break

    case ConfigExitCode.InvalidExport:
      throw new BundleError(
        new Error(
          `The 'config' export in edge function at '${func.path}' must be an object. More on the Edge Functions API at https://ntl.fyi/edge-api.`,
        ),
      )

      break

    case ConfigExitCode.SerializationError:
      throw new BundleError(
        new Error(
          `The 'config' object in the edge function at '${func.path}' must contain primitive values only. More on the Edge Functions API at https://ntl.fyi/edge-api.`,
        ),
      )
      break

    case ConfigExitCode.InvalidDefaultExport:
      throw new BundleError(
        new Error(
          `Default export in '${func.path}' must be a function. More on the Edge Functions API at https://ntl.fyi/edge-api.`,
        ),
      )

    default:
      log.user(`Could not load configuration for edge function at '${func.path}'`)
      log.user(stderr)
  }
}
