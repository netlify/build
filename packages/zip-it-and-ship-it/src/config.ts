import { promises as fs } from 'fs'
import { basename, extname, dirname, join } from 'path'

import isPathInside from 'is-path-inside'
import mergeOptions from 'merge-options'
import { z } from 'zod'

import { FunctionSource } from './function.js'
import { nodeBundler } from './runtimes/node/bundlers/types.js'
import { moduleFormat } from './runtimes/node/utils/module_format.js'
import { minimatch } from './utils/matching.js'

export const functionConfig = z.object({
  externalNodeModules: z.array(z.string()).optional().catch([]),
  generator: z.string().optional().catch(undefined),
  includedFiles: z.array(z.string()).optional().catch([]),
  includedFilesBasePath: z.string().optional().catch(undefined),
  ignoredNodeModules: z.array(z.string()).optional().catch([]),
  name: z.string().optional().catch(undefined),
  nodeBundler: nodeBundler.optional().catch(undefined),
  nodeSourcemap: z.boolean().optional().catch(undefined),
  nodeVersion: z.string().optional().catch(undefined),
  rustTargetDirectory: z.string().optional().catch(undefined),
  schedule: z.string().optional().catch(undefined),
  timeout: z.number().optional().catch(undefined),
  zipGo: z.boolean().optional().catch(undefined),

  // Temporary configuration property, only meant to be used by the deploy
  // configuration API. Once we start emitting ESM files for all ESM functions,
  // we can remove this.
  nodeModuleFormat: moduleFormat.optional().catch(undefined),
})

type FunctionConfig = z.infer<typeof functionConfig>

interface FunctionConfigFile {
  config: FunctionConfig
  version: number
}

type GlobPattern = string

type Config = Record<GlobPattern, FunctionConfig>
type FunctionWithoutConfig = Omit<FunctionSource, 'config'>

const getConfigForFunction = async ({
  config,
  configFileDirectories,
  func,
}: {
  config?: Config
  configFileDirectories?: string[]
  func: FunctionWithoutConfig
}): Promise<FunctionConfig> => {
  const fromConfig = getFromMainConfig({ config, func })

  // We try to read from a function config file if the function directory is
  // inside one of `configFileDirectories`.
  const shouldReadConfigFile = configFileDirectories?.some((directory) => isPathInside(func.mainFile, directory))

  if (!shouldReadConfigFile) {
    return fromConfig
  }

  const fromFile = await getFromFile(func)

  return {
    ...fromConfig,
    ...fromFile,
  }
}

const getFromMainConfig = ({
  config,
  func,
}: {
  config?: Config
  configFileDirectories?: string[]
  func: FunctionWithoutConfig
}): FunctionConfig => {
  if (!config) {
    return {}
  }

  // It's safe to mutate the array because it's local to this function.
  const matches = Object.keys(config)
    .filter((expression) => minimatch(func.name, expression))
    .map((expression) => {
      const wildcardCount = [...expression].filter((char) => char === '*').length

      // The weight increases with the number of hardcoded (i.e. non-wildcard)
      // characters â€” e.g. "netlify" has a higher weight than "net*". We do a
      // subtraction of 1 if there is at least one wildcard character, so that
      // "netlify" has a higher weight than "netlify*".
      const weight = expression.length - wildcardCount - (wildcardCount === 0 ? 0 : 1)

      return {
        expression,
        weight,
      }
    })
    .sort(({ weight: weightA }, { weight: weightB }) => weightA - weightB)
    .map(({ expression }) => config[expression])

  return mergeOptions.apply({ concatArrays: true, ignoreUndefined: true }, matches)
}

const getFromFile = async (func: FunctionWithoutConfig): Promise<FunctionConfig> => {
  const filename = `${basename(func.mainFile, extname(func.mainFile))}.json`
  const configFilePath = join(dirname(func.mainFile), filename)

  try {
    const data = await fs.readFile(configFilePath, 'utf8')
    const configFile = JSON.parse(data) as FunctionConfigFile

    if (configFile.version === 1) {
      return configFile.config
    }
  } catch {
    // no-op
  }

  return {}
}

export { Config, FunctionConfig, FunctionWithoutConfig, getConfigForFunction }
