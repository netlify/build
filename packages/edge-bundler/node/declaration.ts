import { FunctionConfig, HTTPMethod, Path } from './config.js'
import { FeatureFlags } from './feature_flags.js'

interface BaseDeclaration {
  cache?: string
  function: string
  method?: HTTPMethod | HTTPMethod[]
  // todo: remove these two after a while and only support in-source config for non-route related configs
  name?: string
  generator?: string
}

type DeclarationWithPath = BaseDeclaration & {
  path: Path
  excludedPath?: Path | Path[]
}

type DeclarationWithPattern = BaseDeclaration & {
  pattern: string
  excludedPattern?: string | string[]
}

export type Declaration = DeclarationWithPath | DeclarationWithPattern

export const mergeDeclarations = (
  tomlDeclarations: Declaration[],
  userFunctionsConfig: Record<string, FunctionConfig>,
  internalFunctionsConfig: Record<string, FunctionConfig>,
  deployConfigDeclarations: Declaration[],
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _featureFlags: FeatureFlags = {},
) => {
  const functionsVisited: Set<string> = new Set()

  const declarations: Declaration[] = [
    // INTEGRATIONS
    // 1. Declarations from the integrations deploy config
    ...getDeclarationsFromInput(deployConfigDeclarations, internalFunctionsConfig, functionsVisited),
    // 2. Declarations from the integrations ISC
    ...createDeclarationsFromFunctionConfigs(internalFunctionsConfig, functionsVisited),

    // USER
    // 3. Declarations from the users toml config
    ...getDeclarationsFromInput(tomlDeclarations, userFunctionsConfig, functionsVisited),
    // 4. Declarations from the users ISC
    ...createDeclarationsFromFunctionConfigs(userFunctionsConfig, functionsVisited),
  ]

  return declarations
}

const getDeclarationsFromInput = (
  inputDeclarations: Declaration[],
  functionConfigs: Record<string, FunctionConfig>,
  functionsVisited: Set<string>,
): Declaration[] => {
  const declarations: Declaration[] = []
  // For any declaration for which we also have a function configuration object,
  // we replace the path because that object takes precedence.
  for (const declaration of inputDeclarations) {
    const config = functionConfigs[declaration.function]

    if (!config) {
      // If no config is found, add the declaration as is.
      declarations.push(declaration)
    } else if (config.path?.length) {
      // If we have a path specified as either a string or non-empty array,
      // create a declaration for each path.
      const paths = Array.isArray(config.path) ? config.path : [config.path]

      paths.forEach((path) => {
        declarations.push({ ...declaration, cache: config.cache, path })
      })
    } else {
      // With an in-source config without a path, add the config to the declaration.
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { path, excludedPath, ...rest } = config

      declarations.push({ ...declaration, ...rest })
    }

    functionsVisited.add(declaration.function)
  }

  return declarations
}

const createDeclarationsFromFunctionConfigs = (
  functionConfigs: Record<string, FunctionConfig>,
  functionsVisited: Set<string>,
): Declaration[] => {
  const declarations: Declaration[] = []

  for (const name in functionConfigs) {
    const { cache, excludedPath, path, method } = functionConfigs[name]

    // If we have a path specified, create a declaration for each path.
    if (!functionsVisited.has(name) && path) {
      const paths = Array.isArray(path) ? path : [path]

      paths.forEach((singlePath) => {
        const declaration: Declaration = { function: name, path: singlePath }
        if (cache) {
          declaration.cache = cache
        }
        if (method) {
          declaration.method = method
        }
        if (excludedPath) {
          declaration.excludedPath = excludedPath
        }
        declarations.push(declaration)
      })
    }
  }

  return declarations
}

/**
 * Normalizes a regular expression, ensuring it has a leading `^` and trailing
 * `$` characters.
 */
export const normalizePattern = (pattern: string) => {
  let enclosedPattern = pattern

  if (!pattern.startsWith('^')) {
    enclosedPattern = `^${enclosedPattern}`
  }

  if (!pattern.endsWith('$')) {
    enclosedPattern = `${enclosedPattern}$`
  }

  const regexp = new RegExp(enclosedPattern)

  // Strip leading and forward slashes.
  return regexp.toString().slice(1, -1)
}
