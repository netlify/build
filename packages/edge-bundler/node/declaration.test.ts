import { test, expect } from 'vitest'

import { FunctionConfig } from './config.js'
import { Declaration, mergeDeclarations, normalizePattern } from './declaration.js'

const deployConfigDeclarations: Declaration[] = []

test('Ensure the order of edge functions with FF', () => {
  const deployConfigDeclarations: Declaration[] = [
    { function: 'framework-manifest-a', path: '/path1' },
    { function: 'framework-manifest-c', path: '/path3' },
    { function: 'framework-manifest-b', path: '/path2' },
  ]

  const tomlConfig: Declaration[] = [
    { function: 'user-toml-a', path: '/path1' },
    { function: 'user-toml-c', path: '/path3' },
    { function: 'user-toml-b', path: '/path2' },
  ]

  const userFuncConfig = {
    'user-isc-c': { path: ['/path1', '/path2'] },
  } as Record<string, FunctionConfig>

  const internalFuncConfig = {
    'framework-isc-c': { path: ['/path1', '/path2'] },
  } as Record<string, FunctionConfig>

  expect(mergeDeclarations(tomlConfig, userFuncConfig, internalFuncConfig, deployConfigDeclarations)).toMatchSnapshot()
})

test('In-source config takes precedence over netlify.toml config', () => {
  const tomlConfig: Declaration[] = [
    { function: 'geolocation', path: '/geo', cache: 'off' },
    { function: 'json', path: '/json', cache: 'manual' },
  ]

  const userFuncConfig = {
    geolocation: { path: ['/geo-isc', '/*'], cache: 'manual' },
    json: { path: '/json', cache: 'off' },
  } as Record<string, FunctionConfig>

  const expectedDeclarations = [
    { function: 'geolocation', path: '/geo-isc', cache: 'manual' },
    { function: 'geolocation', path: '/*', cache: 'manual' },
    { function: 'json', path: '/json', cache: 'off' },
  ]

  const declarations = mergeDeclarations(tomlConfig, userFuncConfig, {}, deployConfigDeclarations)

  expect(declarations).toEqual(expectedDeclarations)
})

test("Declarations don't break if no in-source config is provided", () => {
  const tomlConfig: Declaration[] = [
    { function: 'geolocation', path: '/geo', cache: 'off' },
    { function: 'json', path: '/json', cache: 'manual' },
  ]

  const userFuncConfig = {
    geolocation: { path: ['/geo-isc'], cache: 'manual' },
    json: {},
  } as Record<string, FunctionConfig>

  const expectedDeclarations = [
    { function: 'geolocation', path: '/geo-isc', cache: 'manual' },
    { function: 'json', path: '/json', cache: 'manual' },
  ]

  const declarations = mergeDeclarations(tomlConfig, userFuncConfig, {}, deployConfigDeclarations)

  expect(declarations).toEqual(expectedDeclarations)
})

test('In-source config works independent of the netlify.toml file if a path is defined and otherwise if no path is set', () => {
  const tomlConfig: Declaration[] = [{ function: 'geolocation', path: '/geo', cache: 'off' }]

  const funcConfigWithPath = {
    json: { path: ['/json', '/json-isc'], cache: 'off' },
  } as Record<string, FunctionConfig>

  const funcConfigWithoutPath = {
    json: { cache: 'off' },
  } as Record<string, FunctionConfig>

  const expectedDeclarationsWithISCPath = [
    { function: 'geolocation', path: '/geo', cache: 'off' },
    { function: 'json', path: '/json', cache: 'off' },
    { function: 'json', path: '/json-isc', cache: 'off' },
  ]

  const expectedDeclarationsWithoutISCPath = [{ function: 'geolocation', path: '/geo', cache: 'off' }]

  const declarationsWithISCPath = mergeDeclarations(tomlConfig, funcConfigWithPath, {}, deployConfigDeclarations)
  expect(declarationsWithISCPath).toEqual(expectedDeclarationsWithISCPath)

  const declarationsWithoutISCPath = mergeDeclarations(tomlConfig, funcConfigWithoutPath, {}, deployConfigDeclarations)
  expect(declarationsWithoutISCPath).toEqual(expectedDeclarationsWithoutISCPath)
})

test('In-source config works if only the cache config property is set', () => {
  const tomlConfig: Declaration[] = [{ function: 'geolocation', path: '/geo', cache: 'off' }]

  const funcConfig = {
    geolocation: { cache: 'manual' },
  } as Record<string, FunctionConfig>

  const expectedDeclarations = [{ function: 'geolocation', path: '/geo', cache: 'manual' }]

  expect(mergeDeclarations(tomlConfig, funcConfig, {}, deployConfigDeclarations)).toEqual(expectedDeclarations)
})

test("In-source config path property works if it's not an array", () => {
  const tomlConfig: Declaration[] = [{ function: 'json', path: '/json-toml', cache: 'off' }]

  const funcConfig = {
    json: { path: '/json', cache: 'manual' },
  } as Record<string, FunctionConfig>

  const expectedDeclarations = [{ function: 'json', path: '/json', cache: 'manual' }]

  expect(mergeDeclarations(tomlConfig, funcConfig, {}, deployConfigDeclarations)).toEqual(expectedDeclarations)
})

test("In-source config path property works if it's not an array and it's not present in toml or deploy config", () => {
  const tomlConfig: Declaration[] = [{ function: 'geolocation', path: '/geo', cache: 'off' }]
  const funcConfig = {
    json: { path: '/json-isc', cache: 'manual' },
  } as Record<string, FunctionConfig>

  const expectedDeclarations = [
    { function: 'geolocation', path: '/geo', cache: 'off' },
    { function: 'json', path: '/json-isc', cache: 'manual' },
  ]

  expect(mergeDeclarations(tomlConfig, funcConfig, {}, deployConfigDeclarations)).toEqual(expectedDeclarations)
})

test('In-source config works if path property is an empty array with cache value specified', () => {
  const tomlConfig: Declaration[] = [{ function: 'json', path: '/json-toml', cache: 'off' }]

  const funcConfig = {
    json: { path: [], cache: 'manual' },
  } as Record<string, FunctionConfig>

  const expectedDeclarations = [{ function: 'json', path: '/json-toml', cache: 'manual' }]

  expect(mergeDeclarations(tomlConfig, funcConfig, {}, deployConfigDeclarations)).toEqual(expectedDeclarations)
})

test('netlify.toml-defined excludedPath are respected', () => {
  const tomlConfig: Declaration[] = [{ function: 'geolocation', path: '/geo/*', excludedPath: '/geo/exclude' }]

  const funcConfig = {}

  const expectedDeclarations = [{ function: 'geolocation', path: '/geo/*', excludedPath: '/geo/exclude' }]

  const declarations = mergeDeclarations(tomlConfig, funcConfig, {}, deployConfigDeclarations)

  expect(declarations).toEqual(expectedDeclarations)
})

test('Does not escape front slashes in a regex pattern if they are already escaped', () => {
  const regexPattern = '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))\\/shows(?:\\/(.*))(.json)?[\\/#\\?]?$'
  const expected = '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))\\/shows(?:\\/(.*))(.json)?[\\/#\\?]?$'

  expect(normalizePattern(regexPattern)).toEqual(expected)
})

test('Escapes front slashes in a regex pattern', () => {
  const regexPattern = '^(?:/(_next/data/[^/]{1,}))?(?:/([^/.]{1,}))/shows(?:/(.*))(.json)?[/#\\?]?$'
  const expected = '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))\\/shows(?:\\/(.*))(.json)?[/#\\?]?$'

  expect(normalizePattern(regexPattern)).toEqual(expected)
})

test('Ensures pattern match on the whole path', () => {
  const regexPattern = '/foo/.*/bar'
  const expected = '^\\/foo\\/.*\\/bar$'

  expect(normalizePattern(regexPattern)).toEqual(expected)
})
