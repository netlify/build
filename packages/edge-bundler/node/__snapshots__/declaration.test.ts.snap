// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ensure the order of edge functions with FF 1`] = `
[
  {
    "function": "framework-manifest-a",
    "path": "/path1",
  },
  {
    "function": "framework-manifest-c",
    "path": "/path3",
  },
  {
    "function": "framework-manifest-b",
    "path": "/path2",
  },
  {
    "function": "framework-isc-c",
    "path": "/path1",
  },
  {
    "function": "framework-isc-c",
    "path": "/path2",
  },
  {
    "function": "user-toml-a",
    "path": "/path1",
  },
  {
    "function": "user-toml-c",
    "path": "/path3",
  },
  {
    "function": "user-toml-b",
    "path": "/path2",
  },
  {
    "function": "user-isc-c",
    "path": "/path1",
  },
  {
    "function": "user-isc-c",
    "path": "/path2",
  },
]
`;
